lib/plugins/expressionParser.js

```diff
@@ -163,6 +165,22 @@ function expressionApplier(context, typeInformation) {
 
         newAttribute.value = applyExpression(attribute.expression, context, typeInformation);
 
+        if (attribute.type === 'Integer') {
+            newAttribute.value = Number.parseInt(newAttribute.value);
+        }
+        else if (attribute.type === 'Float') {
+            newAttribute.value = Number.parseFloat(newAttribute.value);
+        }
+        else if (attribute.type === 'Boolean') {
+            newAttribute.value = (newAttribute.value === 'true');
```

Comment:
Just to be sure... is `'true`' or or shoud it be `true` ?

Reply:
Fixed in 1dfa28969796060f1306123c976449fa9fb9c48c

Reply:
Maybe the test should cover the boolean case (and any other not yet covered now)?

Reply:
@fgalan, I have analyzed in depth the functionality of this plugin to understand its expected results.

As it is explained in [expressionLanguage.md](https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/expressionLanguage.md), the objective of this plugin is: to adapt the information coming from the South Bound APIs to the information reported to the Context Broker by means of expressions.

As you can see in https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/expressionLanguage.md#types, it only supports two types: String or Number.  Also in this part of the documentation, it states: 
>For most arithmetic operations ('*', '/', etc...) if a variable is involved, its value will be cast to  Number, regardless of the original type. If this cast fails (because the value of the variable is not a number, e.g.: 'Fifty'), the overall result will be NaN.

So, if the plugin receives a Boolean, a NULL, a JSON Array or a JSON document, it first tries to convert it to a Number and to apply the expression. 

What happens with NGSIv1: as it only supports strings, for arithmetic expressions, the plugin tries to cast the value to a Number and calculates the expression. If it is not possible to do the conversion, it returns a NAN. If the expression is related to Strings, conversion is not needed. Here it is important to emphasize, that the plugin assumes that all the values that are going to come are strings as you can see in line 136 of https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js:
```
        result = parse(cleanedExpression, context, 'String');
```
The result of the expression will be a Number or a String depending on the type of expression and it is converted back to String in line 151 of this file.

Then, with NGSIv2, we have the following problems:
- Input values for expressions might be not only Strings and Numbers but also instances of the rest of JSON native types.
- Expression nevertheless, only supports Strings and Number.
There, I think that we should add an additional check so that expressions are never executed for unsupported types. With this, some of these lines that we have added would be not needed anymore since JSON arrays and object are not accepted by expressions.

Sorry for the long comment but I wanted to clarify everything before doing more changes.

What do you think?

Reply:
It seems a very comprhensive analysis :) Let me some time to read carefully and think abou it...

Reply:
> Input values for expressions might be not only Strings and Numbers but also instances of the rest of JSON native types.

Not sure if I'm getting the point... input values come from the southbound interface, not related with NGSIv2, and where the type semantics don't exist (as far as I know). All values are strings at southbound. Or maybe I'm wrong?

> Expression nevertheless, only supports Strings and Number.

In the northbound interface, I'd suggest to expand the interpretation of `type` in the following way:

* If type is "String", use the value as it is (no need to cast? is this the current behaviour with the NGSIv1 version of the plugin?)
* If type is "Number", cast the resulting value to JSON number (is this the current behaviour with the NGSIv1 version of the plugin?)
* If type is "Boolean", cast the resulting value to JSON boolean. We need a failover value in the case of casting error, it could be `false`
* If the type is "Null", use `null` as value. This is probably a bit useless, but I'd suggest to support it for compleness.


With regards to Arrays and Object, it would be cool to be able to do someething like this:

```
         "attributes":[
            ...
            {
               "name":"insideHouse",
               "type":"Array",
               "expression": [ "${@temperature}", "${@humidity}", "${@pressure}" ]
            },
            {
               "name":"permiter",
               "type":"Object",
               "expression": {
                 "north": "${@presenceN}",
                 "south": "${@presenceS}",
                 "west": "${@presenceW}",
                 "east": "${@presenceE}"
               }
            }
         ]
```
so, as result of the plugin, if we get a measure like this at southbound

```
t|25|h|79|p|27|pN|0|pS|1|pE|1|pW|0
```

the IOTA sends the following update to CB at northbound:

```
{
   "insideHouse": [ 25, 79, 27],
   "perimeter": {
     "north": 0,
     "south": 1,
     "east": 1,
     "west": 0
}
```

The array and object construct are possible in NGSIv1. Thus, if the current pluging doesn't support (could you confirm?) the above would be out the scope of the migration to NGSIv2. I other words, I'd suggest to focus only in Boolean and Null, but defer Array and Object support to a future enhancemente (and issues in the IOTAgents library repository should be filled about it).

(In fact, regarding Array and Objects, there are interesting challengues to solve... e.g. how to specify type in the sub-elements of the array or Object, i.e. `[ 25, 79, 27 ]` or `[ "25", "79", "27"]`. Maybe the cast could be specified in at expression level? i.e. `"expression": [ "(${@temperature},{Number}" ]`... just thinking aloud :)

The expressionLanguage.md document should be modified in order to take into account the new types and to explain any further NGSIv2 usage consideration to take into acount.


Does this makes any sense to you? I'm not fully sure (you would realize with the several in-line questions between brackets along the discussion above :), as I'm not an expert in this part of the code and the expression plugin is a bit tricky... :)

Reply:
* Input values for expression plugin come from the southbound interface but before passing them as arguments for this plugin, they are converted to NGSI. Just to clarify everything, how is this plugin used?
  * First, in the corresponging IOTA, after calling to iotAgentLib.activate, you have to add this plugin:
```
        iotAgentLib.activate(iotAgentConfig, function() {
            iotAgentLib.clearAll(function() {
                iotAgentLib.addUpdateMiddleware(iotAgentLib.dataPlugins.attributeAlias.update);
                iotAgentLib.addQueryMiddleware(iotAgentLib.dataPlugins.attributeAlias.query);
                iotAgentLib.addUpdateMiddleware(iotAgentLib.dataPlugins.expressionTransformation.update);
                done();
            });
        });
```

  * In line 436 of [ngsiService.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/ngsiService.js), this simply implies that the plugin functionality is added to an array of middlewares to be called when an updateContext request is going to be sent to the CB as a result of having received active attributes from the southbound server.
```
function addUpdateMiddleware(middleware) {
    updateMiddleware.push(middleware);
}
```
  * In line 220 of [ngsiService.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/ngsiService.js), you will find the implementation of sendUpdateValue function.  In line 252 and using a NGSI payload as argument, the update middleware functions are called:
```
        apply(applyMiddlewares, updateMiddleware, payload, typeInformation)
```
* The NGSIv1 version of the plugin does not check any kind of type conversion for the values before applying the expression. It directly considers that the expression has String type: in line 131 of [expressionParser.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js):
```
function processExpression(context) {
    return function(expression) {
        var result,
            cleanedExpression = expression.substr(2, expression.length - 3);

        result = parse(cleanedExpression, context, 'String');

        return {
            original: expression,
            value: result
        };
    };
}
```
* It does not make either any kind of checking to the output values and directly updates the context information to be sent to the CB with the results of the expression. It uses replace instruction to the values will be always a string, which was OK for NGSIv1. In line 151 of the same file:
```
        expressionResult = expressionResult.replace(substitutions[i].original, substitutions[i].value);
```
* What happens then for NGSIv1 version if values which are not strings are received from the soutbound API:
  * If they are numbers: the expression parses seems to calculate correctly the expression and the result is converted to String (see previous point)
  * If they are booleans: the parser directly converts the input value to a number or string depending on the expression type and resolves the expression. 
  * If they are JSON objects or arrays: if the expression is of number type, the parser is not able to do a cast from these types to numbers and it returns NaN. This value is used to update the payload to be sent to the CB. If the expression is a string transformation (i.e., to trim spaces), it converts these values to strings and tries to resolve the operation.

Reply:
So in order to maintain the scope of this PR and just to extend the plugin functionality to NGSIv2, I would follow what you are proposing:
* As happens with NGSIv1, don't do any kind of checking before applying the expressions.
* Try to cast to the appropriate types the results of the expressions. In NGSIv1, this cast was not explicitly (it was done with the replace instruction) needed since values for always represented using Strings. Strings, numbers, booleans and null types will be considered. JSON objects and arrays will be postponed and issues in the repository should be opened to analyze which changes are needed.

I think that we are aligned, aren't we? :)

Reply:
Your proposal makes sense to me. Please, modify code and tests and let's see if all fits together and we have undestand each other. Maybe we don't achieve it in a first iteration, but I have the feeling we are progressing :)

Reply:
Fixed in b217c46b403ab01d6dbbd1290d8e9db4c213e928.

So far, expressions only support the following values: string, integer, float, boolean and null. Therefore, the result of an expression is casted to the corresponding type. JSON arrays and objects are not supported yet so the behaviour will be the same that with NGSIv1.

I have added the following unit tests:
- Update for an integer attribute without expression
- Update for an integer attribute with expression
- Update for a float attribute without expression
- Update for a float attribute with expression
- Update for a Null attribute without expression
- Update for a Null attribute with arithmetic expression
- Update for a Null attribute with string expression (trim spaces)
- Update for a Boolean attribute without expression
- Update for a Boolean attribute with arithmetic expression
- Update for a Boolean attribute with string expression (trim spaces)
- Update for a JSON array without expression
- Update for a JSON object without expression

Reply:
Please find attached coverage report 
![2017-11-29-153631_1366x716_scrot](https://user-images.githubusercontent.com/11388886/33380407-45a70c82-d51b-11e7-8038-3e1330352330.png)
for that function.

Reply:
BTW, unit & lint tests are passed. :)

Reply:
> I have added the following unit tests:
> [...]

I'm trying to map the cases in your bullet list with the ones in expressionBasedTransformations-test.js but I'm not fully sure... Could you annotate with comments in the .js file, e.g. something like this:

```
 describe('When an update comes for expressions with syntax errors', function() {
   // Case: Update for an ..
```

Thanks!

Reply:
Btw, array/object functionality deferred to future issue https://github.com/telefonicaid/iotagent-node-lib/issues/568. Please provide any feedback you think could be usesful there.

Reply:
Fixed in 895d5cc7b47d5fcfd4b81ab3e83a8a858c28c3e7